swifttool/manager.py

```diff
@@ -0,0 +1,288 @@
+                return dev['ip']
+
+    def _calculate_cluster_changes(self):
+        try:
+            self.cluster_config = yaml.load(open(self.def_file, 'r'))
+            expected_config = get_devices(self.cluster_config['zones'])
+            current_config = self._get_all_cluster_devices()
+            # To get devices to be removed we look for devices present in
+            # current_config and not in expected_config
+            # To get list of devices to be added look for devices in expected
+            # not in current
+            if self.scaleup:
+                changes = self._get_devices_changed(expected_config,
+                                                    current_config)
+            else:
+                changes = self._get_devices_changed(current_config,
+                                                    expected_config)
+            self._normalize_changes(changes)
+            # Each ring will be list of devices to be modified
+            # Each device (in ring) will have following attributes:
+            # - zone
+            # - ip
+            # - device
+            # - target (device weight when adding, 0 when removing)
+            # - step (amount to change every time till target reached)
+            # step is defaulted to 10% or -10% of target
+        except Exception as e:
+            print "Error getting device info", e
+            sys.exit(-1)
+
+    def _normalize_changes(self, source):
+        for builder in RING_TYPES:
+            for zone, devadd in source[builder].items():
+                self.cluster_changes.setdefault(builder, [])
+                devices = []
+                for device in devadd:
+                    dev = {}
+                    dev['ip'] = device['ip']
+                    dev['device'] = device['device']
+                    if 'metadata' in device:
+                        dev['metadata'] = device['metadata']
+                    if 'delta' in device:
+                        step = int(math.ceil(device['delta'] /
+                                             self.iterations))
+                    else:
+                        step = int(math.ceil(device['weight'] /
+                                             self.iterations))
+                    if self.scaleup:
+                        dev['target'] = device['weight']
+                        dev['step'] = step
+                    else:
+                        dev['target'] = 0
+                        dev['step'] = int(-1 * step)
+                    devices.append(dev)
+                self.cluster_changes[builder].append({zone: devices})
+
+    @staticmethod
+    def _get_devices_changed(devices1, devices2):
+        devices = {}
+        for builder in RING_TYPES:
+            devices[builder] = {}
+            for zone, devs in devices1[builder].iteritems():
+                if zone not in devices2[builder]:
+                    devices[builder].setdefault(zone, []).extend(devs)
```

Comment:
given your use of `setdefault`, might your needs be met by using a default dict?